pkg/kinflate/app/application.go

```diff
@@ -63,9 +65,28 @@ func New(loader loader.Loader) (Application, error) {
 }
 
 // Resources computes and returns the resources from the manifest.
+// The namehashing for configmap/secrets and resolving name reference is only done
+// in the most top overlay once at the end of getting resources.
 func (a *applicationImpl) Resources() (resource.ResourceCollection, error) {
+	res, err := a.semiResources()
+	if err != nil {
+		return nil, err
+	}
+	t, err := a.getHashAndReferenceTransformer()
+	if err != nil {
+		return nil, err
+	}
+	err = t.Transform(res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
```

Comment:
`semiResources` naming is making things hard to understand. 

I am realizing that now the core logic is not inline with the abstractions we had earlier. Transformers are mixed in a weird way to understand the flow.

I am still trying to form a mental model about the core logic so that it can be simplified. One thing is becoming clear, we need some logic which needs to be applied at the last layer (top most one), for ex. appending hashes to configmap/secrets and then as a result updating name references for those. I hit similar ones while prototyping downward APIs.

Reply:
You are right. The core logic is changed. The transformers are split and put into different parts. 
Appending hashes and name reference are applied only on the top layer. I also feel that we can make this application interface better and more  clear to understand, but haven't got any concrete ideas.